tests/run-make/reproducible-build/rmake.rs

```diff
@@ -0,0 +1,210 @@
+// This test case makes sure that two identical invocations of the compiler
+// (i.e. same code base, same compile-flags, same compiler-versions, etc.)
+// produce the same output. In the past, symbol names of monomorphized functions
+// were not deterministic (which we want to avoid).
+//
+// The test tries to exercise as many different paths into symbol name
+// generation as possible:
+//
+// - regular functions
+// - generic functions
+// - methods
+// - statics
+// - closures
+// - enum variant constructors
+// - tuple struct constructors
+// - drop glue
+// - FnOnce adapters
+// - Trait object shims
+// - Fn Pointer shims
+// See https://github.com/rust-lang/rust/pull/32293
+
+// FIXME(Oneirical): ignore-musl
+// FIXME(Oneirical): two of these test blocks will apparently fail on windows
+// FIXME(Oneirical): try it on test-various
+// # FIXME: Builds of `bin` crate types are not deterministic with debuginfo=2 on
+// # Windows.
+// # See: https://github.com/rust-lang/rust/pull/87320#issuecomment-920105533
+// # Issue: https://github.com/rust-lang/rust/issues/88982
+
+use run_make_support::{bin_name, cwd, diff, rfs, run_in_tmpdir, rust_lib_name, rustc};
+
+fn main() {
+    // Smoke tests. Simple flags, build should be reproducible.
+    smoke_test(None);
+    smoke_test(Some(SmokeFlag::Debug));
+    smoke_test(Some(SmokeFlag::Opt));
+
+    // Builds should be reproducible even through custom library search paths
+    // or remap path prefixes.
+    paths_test(PathsFlag::Link);
+    paths_test(PathsFlag::Remap);
+
+    // Builds should be reproducible even if each build is done in a different directory,
+    // with both --remap-path-prefix and -Z remap-cwd-prefix.
```

Comment:
Suggestion: this does not mention `-C debuginfo=2`, which is significant in the difference between whether binaries are reproducible or not on Linux apparently.

See #89911 for `-C debuginfo=2` problems.